movement/watch_faces/complication/deadline_face.c

```diff
@@ -0,0 +1,649 @@
+            _background_alarm_schedule(settings, context);
+            _change_tick_freq(1, state);
+            state->mode = DEADLINE_FACE_RUNNING;
+            movement_move_to_face(0);
+            break;
+        case EVENT_MODE_BUTTON_UP:
+            _beep_disable(settings);
+            _background_alarm_schedule(settings, context);
+            _running_init(settings, state);
+            _running_display(event, settings, state);
+            state->mode = DEADLINE_FACE_RUNNING;
+            break;
+        case EVENT_BACKGROUND_TASK:
+            _background_alarm_play(settings, state);
+            break;
+        default:
+            return movement_default_loop_handler(event, settings);
+    }
+
+    return true;
+}
+
+/* Setup face */
+void deadline_face_setup(movement_settings_t *settings, uint8_t watch_face_index, void **context_ptr)
+{
+    (void) settings;
+    (void) watch_face_index;
+    if (*context_ptr != NULL)
+        return; /* Skip setup if context available */
+
+    /* Allocate state */
+    *context_ptr = malloc(sizeof(deadline_state_t));
+    memset(*context_ptr, 0, sizeof(deadline_state_t));
+
+    /* Store face index for background tasks */
+    deadline_state_t *state = (deadline_state_t *) *context_ptr;
+    state->face_idx = watch_face_index;
+}
+
+/* Activate face */
+void deadline_face_activate(movement_settings_t *settings, void *context)
+{
+    (void) settings;
+    deadline_state_t *state = (deadline_state_t *) context;
+
+    /* Set display options */
+    _running_init(settings, state);
+    state->mode = DEADLINE_FACE_RUNNING;
+    state->current_index = _closest_deadline(settings, state);
+}
+
+/* Loop face */
+bool deadline_face_loop(movement_event_t event, movement_settings_t *settings, void *context)
+{
+    (void) settings;
+    deadline_state_t *state = (deadline_state_t *) context;
+    switch (state->mode) {
+        case DEADLINE_FACE_SETTING:
+            _setting_loop(event, settings, context);
+            break;
+        default:
+        case DEADLINE_FACE_RUNNING:
+            _running_loop(event, settings, context);
+            break;
+    }
+
+    return true;
+}
+
+/* Resign face */
+void deadline_face_resign(movement_settings_t *settings, void *context)
+{
+    (void) settings;
+    (void) context;
+}
+
+/* Want background task */
+bool deadline_face_wants_background_task(movement_settings_t *settings, void *context)
+{
+    deadline_state_t *state = (deadline_state_t *) context;
+
+    if (!state->alarm_enabled)
+        return false;
+
+    /* Determine closest deadline */
+    watch_date_time now = watch_rtc_get_date_time();
+    uint32_t now_ts = watch_utility_date_time_to_unix_time(now, _get_tz_offset(settings));
+    uint32_t next_ts = state->deadlines[_closest_deadline(settings, state)];
+
+    /* No active deadline */
+    if (next_ts < now_ts)
+        return false;
+
+    /* No deadline within next 60 seconds */
+    if (next_ts >= now_ts + 60)
+        return false;
```

Comment:
Have you considered a simpler implementation that ignores seconds? Then you'd be able to avoid the need for scheduled tasks in the first place. The minutely background task would suffice.

```C
// Filter out the seconds
current_date_time >>= 6;
closest_deadline_date_time >>= 6;

// Handle deadline reached event in loop
return current_date_time == closest_deadline_date_time;
```

It would reduce the resolution of the deadlines from seconds to minutes but I have a feeling that's an acceptable tradeoff. This face is designed to keep track of deadlines spanning years. Timing so precise as to require seconds resolution feels like the domain of countdown timers.

It's up to you. The code seems significantly more sleep mode friendly now and I think it's perfectly acceptable. I will be testing it on my watch soon.

Reply:
I've thought about it, and I'm not really convinced. I usually track deadlines within days, and I don't want to miss the ticking seconds when the last day starts. Moreover, I think that people who use `finetune_face.c` to make their clock as accurate as possible deserve accurate tracking of deadlines.

Reply:
Absolutely. All good points as well.

I'll be testing the new version soon!